core/01_core.js

```diff
@@ -326,43 +328,84 @@
   }
   const InterruptedPrototype = Interrupted.prototype;
 
-  const promiseHooks = {
-    init: [],
-    before: [],
-    after: [],
-    resolve: [],
-    hasBeenSet: false,
-  };
+  const promiseHooks = [
+    [], // init
+    [], // before
+    [], // after
+    [], // resolve
+  ];
 
   function setPromiseHooks(init, before, after, resolve) {
-    if (init) ArrayPrototypePush(promiseHooks.init, init);
-    if (before) ArrayPrototypePush(promiseHooks.before, before);
-    if (after) ArrayPrototypePush(promiseHooks.after, after);
-    if (resolve) ArrayPrototypePush(promiseHooks.resolve, resolve);
+    const hooks = [init, before, after, resolve];
+    for (let i = 0; i < hooks.length; i++) {
+      const hook = hooks[i];
+      // Skip if no callback was provided for this hook type.
+      if (hook == null) {
+        continue;
+      }
+      // Verify that the type of `hook` is a function.
+      if (typeof hook !== "function") {
+        throw new TypeError(`Expected function at position ${i}`);
+      }
+      // Crudely filter out no-op functions.
+      if (
+        RegExpPrototypeTest(/\)(?:\s*|=>){}$/, FunctionPrototypeToString(hook))
+      ) {
+        continue;
+      }
+      // Add the material hook to the list.
+      ArrayPrototypePush(promiseHooks[i], hook);
+    }
 
-    if (!promiseHooks.hasBeenSet) {
-      promiseHooks.hasBeenSet = true;
+    const wrappedHooks = ArrayPrototypeMap(promiseHooks, (hooks) => {
+      switch (hooks.length) {
+        case 0:
+          return noop;
```

Comment:
We should make it `undefined` - with https://github.com/denoland/deno/pull/17604 landed we'll be able to not pass a callback for certain hooks

Reply:
Actually let's wait for https://github.com/denoland/deno/pull/17705

Reply:
yes i was anticipating that

Reply:
OK, done!